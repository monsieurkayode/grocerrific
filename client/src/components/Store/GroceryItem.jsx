/* eslint-disable no-underscore-dangle */
import React from 'react';
import {
  string,
  number,
  func,
  shape,
  arrayOf
} from 'prop-types';

import defaultImage from '../../assets/images/groceries.jpg';

const GroceryItem = ({
  _id,
  name,
  price,
  quantity,
  imageUrl,
  addToCart,
  removeFromCart,
  cartItems
}) => {
  const itemAdded = () => cartItems.findIndex(item => item.id === _id) >= 0;

  const addOrRemove = () => {
    if (!itemAdded()) {
      return addToCart({
        id: _id,
        name,
        price,
        maxQuantity: quantity,
        quantity: 1
      });
    }

    return removeFromCart(_id);
  };

  const buttonText = () => {
    if (quantity === 0) return 'Out of stock';
    return itemAdded()
      ? 'Remove from cart'
      : 'Add to cart';
  };

  return (
    <div className="grocery__list_item">
      <div
        style={{ backgroundImage: `url(${imageUrl})` }}
        className="image__wrapper effect"
      />
      <article className="grocery__info">
        <h3 className="grocery__name">
          {name}<span className="grocery__price">&#8358;{price}</span>
        </h3>
        <div className="grocery__buttons">
          <button
            className="add uppercase"
            type="button"
            onClick={() => addOrRemove()}
            disabled={quantity === 0}
          >
            {buttonText()}
          </button>
        </div>
      </article>
    </div>
  );
};

GroceryItem.defaultProps = {
  imageUrl: defaultImage
};

GroceryItem.propTypes = {
  _id: string.isRequired,
  name: string.isRequired,
  price: number.isRequired,
  quantity: number.isRequired,
  imageUrl: string,
  addToCart: func.isRequired,
  removeFromCart: func.isRequired,
  cartItems: arrayOf(shape({})).isRequired
};

export default GroceryItem;
