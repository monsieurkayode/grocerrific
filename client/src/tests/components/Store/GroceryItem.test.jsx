import React from 'react';
import { shallow } from 'enzyme';
import GroceryItem from '../../../components/Store/GroceryItem';
import mockCartItems from '../../__mocks__/mockCartItems';
import mockGroceryItems from '../../__mocks__/mockGroceries';

const setup = () => {
  const groceryItem = mockGroceryItems[0];

  const props = {
    cartItems: [],
    addToCart: jest.fn(),
    removeFromCart: jest.fn(),
    checkingOut: false,
    ...groceryItem
  };

  const wrapper = shallow(<GroceryItem {...props} />);

  return { wrapper, props };
};

describe('GroceryItem Component', () => {
  const { wrapper, props } = setup();

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render add to cart button', () => {
    const addToCartButton = wrapper.find('.add');
    expect(addToCartButton.length).toBe(1);
  });

  it('should call addToCart', () => {
    const addToCartButton = wrapper.find('.add');
    addToCartButton.simulate('click');
    expect(props.addToCart).toHaveBeenCalled();
  });

  describe('add to cart button should have text', () => {
    it('Add to cart if groceryItem is not yet added to cart', () => {
      const addToCartButton = wrapper.find('.add');
      expect(addToCartButton.text()).toEqual('Add to cart');
    });

    it('Remove from cart if item already added to cart', () => {
      wrapper.setProps({ cartItems: mockCartItems });
      const addToCartButton = wrapper.find('.add');
      expect(addToCartButton.text()).toEqual('Remove from cart');
    });

    it('Out of stock if item quantity is 0', () => {
      wrapper.setProps({ quantity: 0 });
      const addToCartButton = wrapper.find('.add');
      expect(addToCartButton.text()).toEqual('Out of stock');
    });
  });

  it('should call removeFromCart if item already in cart', () => {
    const removeFromCartButton = wrapper.find('.add');
    removeFromCartButton.simulate('click');
    expect(props.removeFromCart).toHaveBeenCalled();
  });
});
