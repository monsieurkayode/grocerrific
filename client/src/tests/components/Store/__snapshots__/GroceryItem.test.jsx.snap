// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroceryItem Component renders correctly 1`] = `
<div
  className="grocery__list_item"
>
  <div
    className="image__wrapper effect"
  >
    <img
      alt=""
      src="test-file-mock"
    />
  </div>
  <article
    className="grocery__info"
  >
    <h3
      className="grocery__name"
    >
      mamador
      <span
        className="grocery__price"
      >
        â‚¦
        1250
      </span>
    </h3>
    <div
      className="grocery__buttons"
    >
      <button
        className="add uppercase"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Add to cart
      </button>
    </div>
  </article>
</div>
`;
